File: app/static/modules/parser.js
const LINE_PATTERN = /^(?:(\d[\d,]*)(?:\s*[xX×:\-–—]\s*|\s+))?(.+?)\s*$/;
const TRAILING_SET_INFO_PATTERN = /\s+\([A-Za-z0-9]{2,6}\)\s+\d+[a-zA-Z]?$/;
const TRAILING_SET_ONLY_PATTERN = /\s+\([A-Za-z0-9]{2,6}\)$/;
const TRAILING_BRACKET_SET_PATTERN = /\s+\[[A-Za-z0-9]{2,8}\]$/;
const TRAILING_CURLY_SET_PATTERN = /\s+\{[A-Za-z0-9]{2,8}\}$/;
const INVALID_QUANTITY_PREFIX_PATTERN = /^(?:[xX]|[+-]\d|\d+\.\d+)\S*\s+/;
const SECTION_HEADER_WITH_COUNT_PATTERN = /^([A-Za-z][A-Za-z\s]+?)\s*\(\d+\)\s*$/;
const GENERIC_SECTION_HEADER_PATTERN = /^([A-Za-z][A-Za-z0-9 '&/+-]{0,80})\s*:\s*$/;
const LEADING_BULLET_PATTERN = /^[\-•–—]\s+/;
const LEADING_CHECKBOX_PATTERN = /^(?:[\-*•–—]\s+)?\[[ xX]\]\s+/;
const TRAILING_ASTERISK_TAG_PATTERN = /\s+\*(?:f|foil|cmdr|commander|maybe|maybeboard|sb|sideboard|mb|mainboard|promo)\*$/i;
const TRAILING_METADATA_LABEL_PATTERN = /\s+(?:\(|\[|\{)(?:foil|etched|showcase|borderless|retro\s+frame|extended\s+art|promo|commander|companion|maybeboard|sideboard|mainboard)(?:\)|\]|\})$/i;
const SECTION_HEADERS = new Set([
  "deck",
  "main",
  "mainboard",
  "maindeck",
  "sideboard",
  "commander",
  "commanders",
  "companions",
  "companion",
  "creatures",
  "spells",
  "lands",
  "artifacts",
  "enchantments",
  "instants",
  "sorceries",
  "planeswalkers",
  "maybeboard",
]);
const SIDEBOARD_MARKERS = new Set(["sideboard", "sb", "side"]);
const COMMANDER_MARKERS = new Set(["commander", "commanders"]);
const COMPANION_MARKERS = new Set(["companion", "companions"]);
const MAYBEBOARD_MARKERS = new Set(["maybeboard"]);

function stripInlineComment(line) {
  const hashIndex = line.indexOf(" #");
  if (hashIndex !== -1) {
    line = line.slice(0, hashIndex);
  }

  const slashIndex = line.indexOf(" // ");
  if (slashIndex !== -1) {
    const remainder = line.slice(slashIndex + 4).trim();
    if (remainder && remainder[0] === remainder[0].toLowerCase()) {
      line = line.slice(0, slashIndex);
    }
  }

  return line.trim();
}

function normalizeCardName(name) {
  let normalized = stripInlineComment(name.trim())
    .replace(LEADING_CHECKBOX_PATTERN, "")
    .replace(LEADING_BULLET_PATTERN, "")
    .trim();

  let previous = null;
  while (previous !== normalized) {
    previous = normalized;
    normalized = normalized
      .replace(TRAILING_SET_INFO_PATTERN, "")
      .replace(TRAILING_SET_ONLY_PATTERN, "")
      .replace(TRAILING_BRACKET_SET_PATTERN, "")
      .replace(TRAILING_CURLY_SET_PATTERN, "")
      .replace(TRAILING_ASTERISK_TAG_PATTERN, "")
      .replace(TRAILING_METADATA_LABEL_PATTERN, "")
      .trim();
  }

  return normalized;
}

export function parseCardLine(line) {
  const trimmed = line.trim().replace(LEADING_CHECKBOX_PATTERN, "").replace(LEADING_BULLET_PATTERN, "");
  if (!trimmed || INVALID_QUANTITY_PREFIX_PATTERN.test(trimmed)) {
    return null;
  }

  const match = trimmed.match(LINE_PATTERN);
  if (!match) {
    return null;
  }

  const quantityText = match[1];
  const quantity = quantityText ? Number.parseInt(quantityText.replaceAll(",", ""), 10) : 1;
  if (!Number.isFinite(quantity) || quantity <= 0) {
    return null;
  }

  const name = normalizeCardName(match[2] || "");
  if (!name) {
    return null;
  }

  return { quantity, name };
}

function isIgnorableDeckLine(line) {
  const stripped = line.trim();
  if (!stripped) return true;
  if (["---", "***", "___"].includes(stripped)) return true;
  if (LEADING_CHECKBOX_PATTERN.test(stripped)) return false;
  return stripped.startsWith("//") || stripped.startsWith("#") || stripped.startsWith("*");
}

function normalizeSectionHeader(line) {
  const lowered = line.toLowerCase().replace(/:+$/, "").trim();
  const match = lowered.match(SECTION_HEADER_WITH_COUNT_PATTERN);
  if (match) {
    return { header: (match[1] || "").trim(), isGenericHeader: true };
  }

  const genericMatch = line.trim().match(GENERIC_SECTION_HEADER_PATTERN);
  if (genericMatch) {
    return { header: (genericMatch[1] || "").trim().toLowerCase(), isGenericHeader: true };
  }

  return { header: lowered, isGenericHeader: false };
}

function lineHasExplicitQuantity(line) {
  const match = line.trim().match(LINE_PATTERN);
  return Boolean(match && match[1]);
}

function shouldIgnoreDeckTitle(lines, index, sawContent) {
  if (sawContent || index !== 0) return false;

  const line = (lines[index] || "").trim();
  if (!line || line.includes(":") || lineHasExplicitQuantity(line)) return false;

  const current = normalizeSectionHeader(line);
  if (current.isGenericHeader || SECTION_HEADERS.has(current.header)) return false;

  for (let i = index + 1; i < lines.length; i += 1) {
    const candidate = (lines[i] || "").trim();
    if (isIgnorableDeckLine(candidate)) continue;
    const future = normalizeSectionHeader(candidate);
    return future.isGenericHeader || SECTION_HEADERS.has(future.header);
  }

  return false;
}

export function parseDeckInputClientDetailed(decklist = "") {
  const details = {
    counts: { mainboard: 0, sideboard: 0, commanders: 0 },
    mainboard: [],
    sideboard: [],
    commanders: [],
    ignored: [],
  };
  let currentSection = "mainboard";
  let sawContent = false;

  const lines = decklist.split(/\r?\n/);

  const recordEntry = (section, parsed, source) => {
    details[section].push({ ...parsed, source });
    details.counts[section] += parsed.quantity;
  };

  lines.forEach((rawLine, index) => {
    const line = rawLine.trim();
    if (isIgnorableDeckLine(line)) return;

    if (shouldIgnoreDeckTitle(lines, index, sawContent)) {
      details.ignored.push({ line, reason: "deck-title" });
      return;
    }

    const { header: lowered, isGenericHeader } = normalizeSectionHeader(line);
    if (SECTION_HEADERS.has(lowered) || isGenericHeader) {
      if (SIDEBOARD_MARKERS.has(lowered) || COMPANION_MARKERS.has(lowered)) {
        currentSection = "sideboard";
      } else if (COMMANDER_MARKERS.has(lowered)) {
        currentSection = "commander";
      } else if (MAYBEBOARD_MARKERS.has(lowered)) {
        currentSection = "ignore";
      } else {
        currentSection = "mainboard";
      }
      sawContent = true;
      return;
    }

    if (["sb:", "sideboard:", "companion:", "companions:"].some((prefix) => lowered.startsWith(prefix))) {
      const parsed = parseCardLine(line.split(":", 2)[1] || "");
      if (parsed) recordEntry("sideboard", parsed, line);
      currentSection = "sideboard";
      sawContent = true;
      return;
    }

    if (["commander:", "commanders:"].some((prefix) => lowered.startsWith(prefix))) {
      const parsed = parseCardLine(line.split(":", 2)[1] || "");
      if (parsed) recordEntry("commanders", parsed, line);
      currentSection = "commander";
      sawContent = true;
      return;
    }

    if (lowered.startsWith("maybeboard:")) {
      details.ignored.push({ line, reason: "inline-maybeboard" });
      return;
    }

    const parsed = parseCardLine(line);
    if (!parsed) return;

    if (currentSection === "ignore") {
      details.ignored.push({ line, reason: "ignored-section" });
      return;
    }

    recordEntry(currentSection, parsed, line);
    sawContent = true;
  });

  return details;
}

export function parseDeckInputClient(decklist = "") {
  return parseDeckInputClientDetailed(decklist).counts;
}

export function parseCommanderFieldCount(text = "") {
  if (!text.trim()) return 0;

  return text
    .replace(/\s+\/\/\s+/g, "\n")
    .replace(/\s+\/\s+/g, "\n")
    .replace(/;/g, "\n")
    .split(/\n+/)
    .map((part) => part.trim())
    .filter(Boolean).length;
}

File: app/services/parser.py
import re


LINE_PATTERN = re.compile(r"^(?:(\d[\d,]*)(?:\s*[xX×:\-–—]\s*|\s+))?(.+?)\s*$")
TRAILING_SET_INFO_PATTERN = re.compile(r"\s+\([A-Za-z0-9]{2,6}\)\s+\d+[a-zA-Z]?$")
TRAILING_SET_ONLY_PATTERN = re.compile(r"\s+\([A-Za-z0-9]{2,6}\)$")
TRAILING_BRACKET_SET_PATTERN = re.compile(r"\s+\[[A-Za-z0-9]{2,8}\]$")
TRAILING_CURLY_SET_PATTERN = re.compile(r"\s+\{[A-Za-z0-9]{2,8}\}$")
INVALID_QUANTITY_PREFIX_PATTERN = re.compile(r"^(?:[xX]|[+-]\d|\d+\.\d+)\S*\s+")
SECTION_HEADER_WITH_COUNT_PATTERN = re.compile(r"^([A-Za-z][A-Za-z\s]+?)\s*\(\d+\)\s*$")
GENERIC_SECTION_HEADER_PATTERN = re.compile(r"^([A-Za-z][A-Za-z0-9 '&/+-]{0,80})\s*:\s*$")
LEADING_BULLET_PATTERN = re.compile(r"^[\-•–—]\s+")
LEADING_CHECKBOX_PATTERN = re.compile(r"^(?:[\-*•–—]\s+)?\[[ xX]\]\s+")
TRAILING_ASTERISK_TAG_PATTERN = re.compile(r"\s+\*(?:f|foil|cmdr|commander|maybe|maybeboard|sb|sideboard|mb|mainboard|promo)\*$", re.IGNORECASE)
TRAILING_METADATA_LABEL_PATTERN = re.compile(
    r"\s+(?:\(|\[|\{)(?:foil|etched|showcase|borderless|retro\s+frame|extended\s+art|promo|commander|companion|maybeboard|sideboard|mainboard)(?:\)|\]|\})$",
    re.IGNORECASE,
)
SECTION_HEADERS = {
    "deck",
    "main",
    "mainboard",
    "maindeck",
    "sideboard",
    "commander",
    "commanders",
    "companions",
    "companion",
    "creatures",
    "spells",
    "lands",
    "artifacts",
    "enchantments",
    "instants",
    "sorceries",
    "planeswalkers",
    "maybeboard",
}
SIDEBOARD_MARKERS = {"sideboard", "sb", "side"}
COMMANDER_MARKERS = {"commander", "commanders"}
COMPANION_MARKERS = {"companion", "companions"}
MAYBEBOARD_MARKERS = {"maybeboard"}


def strip_inline_comment(line: str) -> str:
    hash_index = line.find(" #")
    if hash_index != -1:
        line = line[:hash_index]

    slash_index = line.find(" // ")
    if slash_index != -1:
        remainder = line[slash_index + 4 :].strip()
        if remainder and remainder[:1].islower():
            line = line[:slash_index]

    return line.strip()


def is_ignorable_line(line: str) -> bool:
    stripped = line.strip()
    if not stripped:
        return True
    if stripped in {"---", "***", "___"}:
        return True
    if LEADING_CHECKBOX_PATTERN.match(stripped):
        return False
    if stripped.startswith("//") or stripped.startswith("#") or stripped.startswith("*"):
        return True
    return False


def normalize_name(name: str) -> str:
    name = strip_inline_comment(name.strip())
    name = LEADING_CHECKBOX_PATTERN.sub("", name)
    name = LEADING_BULLET_PATTERN.sub("", name)
    previous = None
    while previous != name:
        previous = name
        name = TRAILING_SET_INFO_PATTERN.sub("", name)
        name = TRAILING_SET_ONLY_PATTERN.sub("", name)
        name = TRAILING_BRACKET_SET_PATTERN.sub("", name)
        name = TRAILING_CURLY_SET_PATTERN.sub("", name)
        name = TRAILING_ASTERISK_TAG_PATTERN.sub("", name)
        name = TRAILING_METADATA_LABEL_PATTERN.sub("", name)
    return name.strip()


def normalize_section_header(line: str) -> tuple[str, bool]:
    lowered = line.lower().rstrip(":").strip()
    match = SECTION_HEADER_WITH_COUNT_PATTERN.match(lowered)
    if match:
        return match.group(1).strip(), True

    generic_match = GENERIC_SECTION_HEADER_PATTERN.match(line.strip())
    if generic_match:
        return generic_match.group(1).strip().lower(), True

    return lowered, False


def line_has_explicit_quantity(line: str) -> bool:
    match = LINE_PATTERN.match(line.strip())
    return bool(match and match.group(1))


def should_ignore_deck_title(lines: list[str], index: int, saw_content: bool) -> bool:
    if saw_content or index != 0:
        return False

    line = lines[index].strip()
    if not line or ":" in line or line_has_explicit_quantity(line):
        return False

    normalized, is_header = normalize_section_header(line)
    if is_header or normalized in SECTION_HEADERS:
        return False

    for future_line in lines[index + 1 :]:
        candidate = future_line.strip()
        if is_ignorable_line(candidate):
            continue
        future_normalized, future_is_header = normalize_section_header(candidate)
        return future_is_header or future_normalized in SECTION_HEADERS

    return False


def parse_card_line(line: str) -> dict | None:
    cleaned_line = LEADING_CHECKBOX_PATTERN.sub("", line.strip())
    cleaned_line = LEADING_BULLET_PATTERN.sub("", cleaned_line)

    if INVALID_QUANTITY_PREFIX_PATTERN.match(cleaned_line):
        return None

    match = LINE_PATTERN.match(cleaned_line)
    if not match:
        return None

    quantity_text = match.group(1)
    quantity = int(quantity_text.replace(",", "")) if quantity_text else 1
    if quantity <= 0:
        return None

    name = normalize_name(match.group(2))
    if not name:
        return None

    return {"quantity": quantity, "name": name}


def parse_deck_input(decklist: str) -> tuple[list[dict], list[dict], list[dict]]:
    mainboard: list[dict] = []
    sideboard: list[dict] = []
    commanders: list[dict] = []
    current_section = "mainboard"
    lines = decklist.splitlines()
    saw_content = False

    for index, raw_line in enumerate(lines):
        line = raw_line.strip()
        if is_ignorable_line(line):
            continue

        if should_ignore_deck_title(lines, index, saw_content):
            continue

        lowered, is_generic_header = normalize_section_header(line)
        if lowered in SECTION_HEADERS or is_generic_header:
            if lowered in SIDEBOARD_MARKERS or lowered in COMPANION_MARKERS:
                current_section = "sideboard"
            elif lowered in COMMANDER_MARKERS:
                current_section = "commander"
            elif lowered in MAYBEBOARD_MARKERS:
                current_section = "ignore"
            else:
                current_section = "mainboard"
            saw_content = True
            continue

        if (
            lowered.startswith("sb:")
            or lowered.startswith("sideboard:")
            or lowered.startswith("companion:")
            or lowered.startswith("companions:")
        ):
            card_line = line.split(":", 1)[1].strip()
            parsed = parse_card_line(card_line)
            if parsed:
                sideboard.append(parsed)
            current_section = "sideboard"
            saw_content = True
            continue

        if lowered.startswith("commander:") or lowered.startswith("commanders:"):
            card_line = line.split(":", 1)[1].strip()
            parsed = parse_card_line(card_line)
            if parsed:
                commanders.append(parsed)
            current_section = "commander"
            saw_content = True
            continue

        if lowered.startswith("maybeboard:"):
            continue

        parsed = parse_card_line(line)
        if not parsed:
            continue

        if current_section == "sideboard":
            sideboard.append(parsed)
        elif current_section == "commander":
            commanders.append(parsed)
        elif current_section == "ignore":
            continue
        else:
            mainboard.append(parsed)
        saw_content = True

    return mainboard, sideboard, commanders


def parse_deck_sections(decklist: str) -> tuple[list[dict], list[dict]]:
    mainboard, sideboard, _ = parse_deck_input(decklist)
    return mainboard, sideboard


def parse_decklist(decklist: str) -> list[dict]:
    mainboard, _, _ = parse_deck_input(decklist)
    return mainboard
